### ISSUE TITLE
Whisper is not learning a new tokenizer, even when i make test and train dataset the same

### ISSUE THREAD

**Issue by P-Sood** [huggingface/transformers#27583]
### System Info

- `transformers` version: 4.35.2
- Platform: Linux-5.15.120+-x86_64-with-glibc2.35
- Python version: 3.10.12
- Huggingface_hub version: 0.19.3
- Safetensors version: 0.4.0
- Accelerate version: 0.24.1
- Accelerate config: 	not found
- PyTorch version (GPU?): 2.1.0+cu118 (False)
- Tensorflow version (GPU?): 2.14.0 (False)
- Flax version (CPU?/GPU?/TPU?): 0.7.5 (cpu)
- Jax version: 0.4.20
- JaxLib version: 0.4.20
- Using GPU in script?: Yes
- Using distributed or parallel set-up in script?: No

### Who can help?

@sanchit-gandhi 

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

Hello, I want to take the audio at my workplace and transform it into a transcription; however, with base whisper, it seems as though it isn't that good. So, I have been wanting to create my own tokenizer that can understand jargon and output that jargon better. Stuff similar to acronyms.  Below I have shown my steps in 
1) Creating Tokenizer
2) Preprocessing data pipeline
3) Model init, and configuration
4) Model outputs

I run this using huggingface trainer, with the generate option. Is it my data size? i have scoured online to try and find some sort of solution but they all just say it works. I am at my wits end and would appreciate any help on getting this tokenizer to learn my jargon. 

Thank you in advance :)

## Creating the tokenizer
```python
from tokenizers import Tokenizer, models, pre_tokenizers, decoders, trainers
from transformers import WhisperTokenizer

# Initialize a tokenizer
tokenizer = Tokenizer(models.BPE())

# Pre-tokenizer responsible for converting the text to a stream of characters
tokenizer.pre_tokenizer = pre_tokenizers.Whitespace()#ByteLevel(add_prefix_space=False)

# Decoder responsible for converting the tokens back to a string
tokenizer.decoder = decoders.ByteLevel()

# Trainer responsible for training the BPE model
tokenizer.trainers = trainers.BpeTrainer(vocab_size=1000, min_frequency=2 , special_tokens=spec_tok)

# Training the tokenizer
tokenizer.train(["file.txt"])

# Save the tokenizer
tokenizer.save("NewWhisperTokenizer.json")

f = open('NewWhisperTokenizer.json')

# returns JSON object as
# a dictionary
data = json.load(f)
with open("vocab.json", "w") as outfile:
    json.dump(data['model']['vocab'], outfile)
with open("merges.txt", "w") as outfile:
    json.dump(data['model']['merges'], outfile)


tokenizer = WhisperTokenizer("vocab.json", "merges.txt" , errors = "replace", unk_token = "<|endoftext|>", bos_token = "<|endoftext|>", eos_token = "<|endoftext|>", pad_token = "<|endoftext|>")
tokenizer.add_special_tokens(WhisperTokenizer.from_pretrained("openai/whisper-tiny").special_tokens_map_extended)
tokenizer.save_pretrained("new_tok")
```

`len(tokenizer) == 193`

## Preprocessing steps
```python
def prepare_dataset(batch):
    audio = batch["audio"]
    batch["input_features"] = feature_extractor(audio["array"], sampling_rate=audio["sampling_rate"]).input_features[0]
    temp_labels = tokenizer(batch["phonetic_detail"]["utterance"]).input_ids
    batch["label"] = [label for sentence_labels in temp_labels for label in sentence_labels]
    return batch

@dataclass
class DataCollatorSpeechSeq2SeqWithPadding:
    tokenizer: Any
    feature_extractor: Any
    def __call__(self, features: List[Dict[str, Union[List[int], torch.Tensor]]]) -> Dict[str, torch.Tensor]:
        input_features = [{"input_features": feature["input_features"]} for feature in features]
        batch = self.feature_extractor.pad(input_features, return_tensors="pt")
        label_features = [{"input_ids": feature["label"]} for feature in features]
        labels_batch = self.tokenizer.pad(label_features, return_tensors="pt")


        labels = labels_batch["input_ids"].masked_fill(labels_batch.attention_mask.ne(1), -100)
        if (labels[:, 0] == self.tokenizer.bos_token_id).all().cpu().item():
            labels = labels[:, 1:]
        batch["labels"] = labels
        return batch

data_collator = DataCollatorSpeechSeq2SeqWithPadding(tokenizer , feature_extractor)
```

`len(train_dataset) == 4000`
`len(test_dataset) == 1000`

## Model Config
```python
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

voc = tokenizer.get_vocab()

model_Gen = WhisperForConditionalGeneration.from_pretrained("openai/whisper-tiny")
model_Gen = model_Gen.to(device)

model_Gen.resize_token_embeddings(len(tokenizer))

model_Gen.config.pad_token_id = tokenizer.pad_token_id
model_Gen.config.decoder_start_token_id = voc['<|startoftranscript|>']
model_Gen.config.eos_token_id = tokenizer.eos_token_id
model_Gen.config.bos_token_id = tokenizer.bos_token_id
model_Gen.config.suppress_tokens = []
model_Gen.config.forced_decoder_ids = None
model_Gen.config.begin_suppress_tokens = [
    tokenizer.pad_token_id
  ]

model_Gen.generation_config.pad_token_id = tokenizer.pad_token_id
model_Gen.generation_config.decoder_start_token_id = voc['<|startoftranscript|>']
model_Gen.generation_config.eos_token_id = tokenizer.eos_token_id
model_Gen.generation_config.bos_token_id = tokenizer.bos_token_id
model_Gen.generation_config.suppress_tokens = []
model_Gen.generation_config.forced_decoder_ids = None
model_Gen.generation_config.begin_suppress_tokens = [
    tokenizer.pad_token_id
  ]

model_Gen.generation_config.no_timestamps_token_id = voc['<|notimestamps|>']
```

## Huggingface Trainer
Here I have made the dataset the same 30 examples to see if it would give me complete overprediction, but even with setting train and test to be the same, it is not overfitting at all.
```python
training_args = Seq2SeqTrainingArguments(
  output_dir='training_output',
  logging_dir='./logs',
  group_by_length=True,
  per_device_train_batch_size=1,
  gradient_accumulation_steps=1,  # increase by 2x for every 2x decrease in batch size
  per_device_eval_batch_size=1,
  num_train_epochs=8,
  gradient_checkpointing=True,
  lr_scheduler_type = "cosine_with_restarts",
  save_strategy='epoch',
  evaluation_strategy='epoch',
  logging_strategy='epoch',
  learning_rate=1e-2,
  weight_decay=0.005,
  # warmup_steps=36,
  save_total_limit=4,
  push_to_hub=False,
  predict_with_generate=True,
  generation_max_length=225,
  load_best_model_at_end=True,
  greater_is_better=False,
  generation_num_beams = 4,
  # fp16 = True,

  report_to="wandb", # Turn this off for pdb debug

)

trainer = CustomTrainer(
    compute_metrics=compute_metrics,
    args=training_args,
    model=model_Gen,
    data_collator=data_collator,
    tokenizer=processor.feature_extractor,
    train_dataset=new_test['test'] ,
    eval_dataset=new_test['test'],
)

trainer.evaluate()
```

## Outputs after second epoch

```python
tokenizer.batch_decode(pred.predictions , skip_special_tokens = True)
['', '', 'uwuwuwuwuwuwuwuwuwuwuwuwuwuwuwuwuwuwuwuwuwuwuwuwuwuwuwuwuwuwuwuwuwuwuwuwuwuwuwuwuw', 'k', '', 'k', 'kkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkk', 
'awawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawaw', 'awawawaw', '', '', '', 'jjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjj', '', 'jjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjj', 'uweuwuwuwuwuwuwuwuwuwuwuwuwuwuwuwuwuwuwuwuwuwuwuwuwuwuwuwuwuwuwuwuwuwuwuwuwuwuwuwuwuwuwuwuwuwuwuwuwuwuwuwuwuwuwuwuwuwuwuwuwuw', '', 
'axaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxaxax', '', 
'kuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhkuhk', 
'eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee', 
'eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee',
 'awawawaw', 
'eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee', 
'awawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawaw',
 '', 
'jjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjjj']

```

### Expected behavior

More understandable text descriptions

**Comment by ArthurZucker** [comment]
Hey 🤗 thanks a lot for opening an issue and using transformers! 

We try to keep the github issues for bugs/feature requests.
Could you ask your question on the [forum](https://discuss.huggingface.co/) instead? I'm sure the community will be of help!

Otherwise you should follow the tutorial ressources on how to train a whisper model see:
- https://github.com/huggingface/distil-whisper
- https://discuss.huggingface.co/t/fine-tuning-whisper-on-my-own-dataset-with-a-customized-tokenizer/25903

Thanks!

**Comment by P-Sood** [comment]
Hello @ArthurZucker I shall post it on the huggingface forums as you request. 

I saw that second post with training on the custom tokenizer. However, the fix they used was to switch it back to the regular pretrained tokenizer and just train for longer. So that doesn't seem like it would have too much effect on me. 

The other issue I looked at [here](https://github.com/huggingface/transformers/issues/25503) was on the huggingface bugs page so I decided to post it here as well.

They also had a similar issue, but they needed help to get the model to train, and had no information on the results after the code was correct. Maybe I should leave a comment at the author of that issue, seeing if he got it work.

Anyways, thanks for the info, ill post it on the forums.

**Comment by ArthurZucker** [comment]
I am not sure why you need to train a new tokenizer but I don't recommend it. You are completely losing the mapping from input_ids and tokens, thus the preptrained model is rendered useless. You should add tokens to the tokenizers rather than train a new one from scratch if you want to leverage the pretrained checkpoint

**Comment by sanchit-gandhi** [comment]
Do you know ahead of time what the kind of jargon is? You could first try Whisper prompting by putting your 'jargon' as the prompt:
```python
processor = WhisperProcessor.from_pretrained("openai/whisper-tiny")
model = WhisperForConditionalGeneration.from_pretrained("openai/whisper-tiny")
input_features = processor(input_speech, return_tensors="pt").input_features

# --- Without prompt ---
prompt_ids = processor.get_prompt_ids("Leighton")
output_without_prompt = model.generate(input_features)
print(processor.decode(output_without_prompt[0]))
# "<|startoftranscript|><|en|><|transcribe|><|notimestamps|> He has grave doubts whether Sir Frederick Layton's work is really Greek after all and can discover in it but little of Rocky Ithaca.<|endoftext|>"

# --- With prompt ---
prompt_ids = processor.get_prompt_ids("Leighton")
output_with_prompt = model.generate(input_features, prompt_ids=prompt_ids)
print(processor.decode(output_with_prompt[0]))
# "<|startofprev|> Leighton<|startoftranscript|><|en|><|transcribe|><|notimestamps|> He has grave doubts whether Sir Frederick Leighton's work is really Greek after all and can discover in it but little of Rocky Ithaca.<|endoftext|>"
```

Your next best method would be fine-tuning using the original tokenizer on your dataset, using as much data as possible: https://huggingface.co/blog/fine-tune-whisper

If you're in a low-data regime, freezing the encoder is recommended. Call this line before you do `trainer.train()`:
```
model.freeze_encoder()
```

After that, see this issue for recommendations for custom vocabulary: https://discuss.huggingface.co/t/adding-custom-vocabularies-on-whisper/29311?u=nbroad. Note that this will require **more** data than standard fine-tuning, so you should be completely sure standard fine-tuning with the original tokenizer doesn't work before trying this. Also note that as @ArthurZucker mentioned, it is not recommended to completely reset the tokenizer, but rather append the new vocabulary to the tokenizer.

**Comment by hilevon** [comment]
Hi @ArthurZucker, @sanchit-gandhi

The whisper tokenizer does not work "as expected" for armenian language - instead of getting just a few tokens (having a few letters) I got more that 20 tokens (having just 1-2 letters).

# Text sample
english: "I wanna dance"
armenian: "Ես ուզում եմ պարել"

The  armenian language is officially supported by whisper.
Instead of getting just a few tokens I got 22 tokens.
As a comparison I've tokenized an english sample and got just 3 tokens.

# Code sample
from transformers import WhisperTokenizer

tokenizer = WhisperTokenizer.from_pretrained("openai/whisper-base")

tokens_en = tokenizer.tokenize("I wanna dance")
print(tokens_en) 
# ['I', 'Ġwanna', 'Ġdance']

tokens_hy = tokenizer.tokenize("Ես ուզում եմ պարել") 
print(tokens_hy)
# ['Ô', 'µ', 'Õ', '½', 'ĠÕ', '¸', 'ÖĤ', 'Õ', '¦', 'Õ¸ÖĤ', 'Õ', '´', 'ĠÕ', '¥', 'Õ', '´', 'ĠÕ', 'º', 'Õ¡', 'ÖĢ', 'Õ¥Õ', '¬']

I tried booth WhisperTokenizer and WhisperTokenizerFast, but got the same result.
So, the question is how can I train the WhisperTokenizer on my own armenian text corpus to improve tokenization?
Thank you.

**Comment by ArthurZucker** [comment]
Hey! I would recommend you to use `tokenizer.train_new_from_iterator` for example! https://huggingface.co/learn/nlp-course/en/chapter6/2 for more details

**Comment by ArthurZucker** [comment]
The issue is that you might have to train the model as well, which is much more compilcated

**Comment by hilevon** [comment]
@ArthurZucker Thank you for your response.

I've used huggingface great tutorial and could train the tokenizer on armenian text corpus using old_tokenizer.train_new_from_iterator().

## Here is a result:
sentence = "Ես ուզում եմ պարել"
print(old_tokenizer.tokenize(sentence))
print(new_tokenizer.tokenize(sentence))

#### ['Ô', 'µ', 'Õ', '½', 'ĠÕ', '¸', 'ÖĤ', 'Õ', '¦', 'Õ¸ÖĤ', 'Õ', '´', 'ĠÕ', '¥', 'Õ', '´', 'ĠÕ', 'º', 'Õ¡', 'ÖĢ', 'Õ¥Õ', '¬']
#### ['ÔµÕ½', 'ĠÕ¸ÖĤÕ¦Õ¸ÖĤÕ´', 'ĠÕ¥Õ´', 'ĠÕºÕ¡ÖĢÕ¥Õ¬']

It seems now the text is tokenized better.

However when I used the new tokenizer for prediction, I got an empty prediction. I used pipeline. Below is my code:
pipe = pipeline(
        "automatic-speech-recognition",
        model=model,
        # tokenizer=processor.tokenizer,
        tokenizer=new_tokenizer,
        feature_extractor=processor.feature_extractor,
        max_new_tokens=128,
        chunk_length_s=30,
        batch_size=16,
        device=device,
)
predicted = pipe(
        audio_path,
        generate_kwargs = {
                "task":"transcribe", 
                "language":"<|hy|>",
        }
)
print(predicted)
#### {'text': ''}

## I found
Then I used new_tokenizer.save_pretrained() to save trained tokenizer. I found that the vocab.json vocabularies are different for old and new tokenizers - special tokens have different ids:
old tokenizer: "<|endoftext|>": 50257
new_tokenizer: "<|endoftext|>": 0

And vocab of new tokenizer also has extra special tokens like:
"<|startoftranscript|>": 1,
"<|en|>": 2,
"<|zh|>": 3,

## Questions
### 1. Is it possible to train a tokenizer by using tokenizer.train_new_from_iterator() and avoid model training?
### 2. Why I got an empty prediction with the new tokenizer?
### 3. Having "<|endoftext|>" special token id 50257 and 0 for old and new tokenizers, is it okey? Or I'm doing something wrong?
### 4. Having extra about 100 special tokens in new tokenizer's vocabulary is it okey?

Thank you!

**Comment by LevonHakhoyan** [comment]
@ArthurZucker Thank you for the detailed explanation.

It seems there is an issue when we want to train GPT2 tokenizer using tokenizer.train_new_from_iterator() where tokenizer
is "tokenizer=GPT2TokenizerFast.from_pretrained("gpt2")".

As I understand the "<|endoftext|>" special toke id must be the last one (or one of the last ones if other special tokens are used as well) in vocab. For example in case of vocab size 50258 the id is 50257. The issue is that after tokenizer training it's is 0.

Below are two different approaches to train the tokenizer:
1th approach - we get a valid tokenzier ( we created a BPE tokenizer from scratch and trained it).
2nd approach -  the tokenizer is invalid (we used GPT2TokenizerFast.from_pretrained("gpt2")).

## 1th approache - valid tokenizer
```
from tokenizers import (
    decoders,
    models,
    normalizers,
    pre_tokenizers,
    processors,
    trainers,
    Tokenizer,
)
from datasets import load_dataset
from transformers import GPT2TokenizerFast

raw_datasets = load_dataset("wikimedia/wikipedia", "20231101.hy")
raw_datasets_train = raw_datasets["train"].select(range(10000))

def get_training_corpus():
    return (
        raw_datasets_train[i : i + 1000]["text"]
        for i in range(0, len(raw_datasets_train), 1000)
    )

tokenizer = Tokenizer(models.BPE())
tokenizer.pre_tokenizer = pre_tokenizers.ByteLevel(add_prefix_space=False)

vocab_size = 50257
special_tokens = ["<|endoftext|>"]

trainer = trainers.BpeTrainer(vocab_size=vocab_size)
tokenizer.add_special_tokens(special_tokens)
tokenizer.add_tokens(special_tokens)
tokenizer.train_from_iterator(get_training_corpus(), trainer=trainer)

tokenizer.post_processor = processors.ByteLevel(trim_offsets=False)

wrapped_tokenizer = GPT2TokenizerFast(tokenizer_object=tokenizer,
    bos_token='<|startoftext|>',
    eos_token='<|endoftext|>',
    unk_token='<|unknown|>',
    pad_token='<|pad|>'
)

print(tokenizer.get_vocab()["<|endoftext|>"])                      # =========> 0
print(wrapped_tokenizer.get_vocab()["<|endoftext|>"])     # =========> 50257
print(len(tokenizer.get_vocab()))                                           # =========> 50258
```

## 2nd approaches - invalid tokenizer
```
old_tokenizer = GPT2TokenizerFast.from_pretrained("gpt2")
new_tokenizer = old_tokenizer.train_new_from_iterator(get_training_corpus(), vocab_size)

print(old_tokenizer.get_vocab()["<|endoftext|>"])   # =========> 50256
print(new_tokenizer.get_vocab()["<|endoftext|>"])   # =========> 0
```

**Comment by ArthurZucker** [comment]
> As I understand the "<|endoftext|>" special toke id must be the last one (or one of the last ones if other special tokens are used as well) in vocab.

This assumption is not necessarily true. The most important thing is that it stays at the same position if you want to re-use the tokenizer. 

Now, when training the tokenizer, you don't need the special token. So you should add it afterwards or, give {"<endoftext>":58200}. as the initial vocab to your tokenizer. 

Another thing is, you should not use the ByteLevel pretokenizer but the normalizer. 
If you try to decoder 0, you will see that it will not be `"<|endoftext|>"` 😉

**Comment by ArthurZucker** [comment]
GYI @itazap

**Comment by LevonHakhoyan** [comment]
@ArthurZucker Thank you for the clarification.

So, could you please provide me some python code to train GPT2 tokenizer and model, as well as some inference code?
I used Hugging Face NLP course to train the GPT2 tokenizer. Then I used it to fine tune a GPT2 model. And finally when I tried to make some inference it totally failed. That's why I assume there is some issue with "tokenizer.train_new_from_iterator()" function for GPT2 tokenizer.

**Comment by ArthurZucker** [comment]
Honestly it's a bit complicated 😅 
TLDR: 
- if you train a new tokenizer entirely, you are doomed
- if you train from an olde one, then okay some tokens will match, but some will not -> some words will be completely unsee by the model 
- if you resize the embedding without the proper function, you are doomed as well

**Comment by LevonHakhoyan** [comment]
Yes, it is 😅
After a lot of research, I could successfully trained a new tokenizer on new data.
I've also fine-tuned the model, and it works perfectly.

**Comment by ArthurZucker** [comment]
Would you mind sharing what unblocked you?! 🤗 I am super curious
